test-data/unit/check-based-infer-function-types.test

```diff
@@ -0,0 +1,161 @@
+reveal_type(B2.__new__)  # N: Revealed type is "def (cls: Type[__main__.B2], a: builtins.int) -> __main__.B2"
+reveal_type(A2.__init__)  # N: Revealed type is "def (self: __main__.A2, a: builtins.int)"
+reveal_type(B2.__init__)  # N: Revealed type is "def (self: __main__.B2, a: builtins.int)"
+
+
+[case testDontInferFunctionTypes]
+# flags: --no-infer-function-types --allow-untyped-defs --allow-incomplete-defs --allow-any-expr --no-default-return
+
+def f(): ...
+reveal_type(f)  # N: Revealed type is "def () -> Any"
+
+def g(i=1, b=""): ...
+reveal_type(g)  # N: Revealed type is "def (i: Any =, b: Any =) -> Any"
+
+class A1:
+    def __new__(cls): ...
+    def __init__(self): ...
+class B1(A1):
+    def __new__(cls): ...
+    def __init__(self): ...
+
+reveal_type(A1.__new__)  # N: Revealed type is "def (cls: Type[__main__.A1]) -> Any"
+reveal_type(B1.__new__)  # N: Revealed type is "def (cls: Type[__main__.B1]) -> Any"
+reveal_type(A1.__init__)  # N: Revealed type is "def (self: __main__.A1) -> Any"
+reveal_type(B1.__init__)  # N: Revealed type is "def (self: __main__.B1) -> Any"
+
+class A2:
+    def __new__(cls, a: int): ...
+    def __init__(self, a: int): ...
+class B2(A2):
+    def __new__(cls, a): ...
+    def __init__(self, a): ...
+
+reveal_type(A2.__new__)  # N: Revealed type is "def (cls: Type[__main__.A2], a: builtins.int) -> Any"
+reveal_type(B2.__new__)  # N: Revealed type is "def (cls: Type[__main__.B2], a: Any) -> Any"
+reveal_type(A2.__init__)  # N: Revealed type is "def (self: __main__.A2, a: builtins.int)"
+reveal_type(B2.__init__)  # N: Revealed type is "def (self: __main__.B2, a: Any) -> Any"
+
+
+[case testSimpleOverload]
+from typing import overload
+
+@overload
+def f(i: int) -> int: ...
+@overload
+def f(i: str) -> str: ...
+
+def f(i):
+    reveal_type(i)  # N: Revealed type is "builtins.int | builtins.str"
+    reveal_type(f)  # N: Revealed type is "Overload(def (i: builtins.int) -> builtins.int, def (i: builtins.str) -> builtins.str)"
+    if i:
+        return "asdf"
+    elif i:
+        return 100
+    return None  # E: Incompatible return value type (got "None", expected "Union[int, str]")  [return-value]
+
+
+[case testPartialOverload]
+from typing import overload
+
+@overload
+def f(*, i: int, j: int) -> int: ...
+@overload
+def f(*, j: str) -> str: ...
+
+def f(i: object = 1, j = "1"):
+    reveal_type(i)  # N: Revealed type is "builtins.object"
+    reveal_type(j)  # N: Revealed type is "builtins.int | builtins.str"
+    if j:
+        return "asdf"
+    elif j:
+        return 100
+    return None  # E: Incompatible return value type (got "None", expected "Union[int, str]")  [return-value]
+
+
+[case testVariadic]
+# flags: --allow-any-expr --disable-error-code truthy-bool --implicit-optional
+from typing import overload
+
+@overload
+def f(__a: str): ...
+@overload
+def f(__a: int): ...
+
+def f(*args):  # E: Function is missing a type annotation for one or more arguments  [no-untyped-def]
```

Comment:
shouldnt it infer them? or is the error that the impl doesnt match the overloads? if so should the error message say so?

Reply:
It gets much more complicated when it's inferring star args, so I am just ignoring for now.

Reply:
is there an issue/TODO/xfail to fix it? or maybe make the error more specific to say that its not supported rather than that the type is wrong

Reply:
I'll raise a follow up issue with the improvements.

I don't really care about makeing a specific error for this if it's planned to be upgraded later, I'll just document it.

Reply:
#291